File: packages/go/luna/client_test.go
package luna_test

import (
	"testing"

	"github.com/eclipse-softworks/luna-sdk-go/luna"
	"github.com/eclipse-softworks/luna-sdk-go/luna/errors"
)

const validAPIKey = "lk_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

func TestNewClient(t *testing.T) {
	t.Run("creates client with API key", func(t *testing.T) {
		client := luna.NewClient(luna.WithAPIKey(validAPIKey))
		if client == nil {
			t.Fatal("expected client to be created")
		}
	})

	t.Run("creates client with tokens", func(t *testing.T) {
		client := luna.NewClient(luna.WithTokens("access-token", "refresh-token"))
		if client == nil {
			t.Fatal("expected client to be created")
		}
	})

	t.Run("panics without auth", func(t *testing.T) {
		defer func() {
			if r := recover(); r == nil {
				t.Fatal("expected panic")
			}
		}()
		luna.NewClient()
	})

	t.Run("accepts custom base URL", func(t *testing.T) {
		client := luna.NewClient(
			luna.WithAPIKey(validAPIKey),
			luna.WithBaseURL("https://api.staging.eclipse.dev"),
		)
		if client == nil {
			t.Fatal("expected client to be created")
		}
	})

	t.Run("accepts custom timeout", func(t *testing.T) {
		client := luna.NewClient(
			luna.WithAPIKey(validAPIKey),
			luna.WithTimeout(60000),
		)
		if client == nil {
			t.Fatal("expected client to be created")
		}
	})

	t.Run("accepts custom max retries", func(t *testing.T) {
		client := luna.NewClient(
			luna.WithAPIKey(validAPIKey),
			luna.WithMaxRetries(5),
		)
		if client == nil {
			t.Fatal("expected client to be created")
		}
	})
}

func TestClientResources(t *testing.T) {
	client := luna.NewClient(luna.WithAPIKey(validAPIKey))

	t.Run("exposes users resource", func(t *testing.T) {
		if client.Users() == nil {
			t.Fatal("expected users resource")
		}
	})

	t.Run("exposes projects resource", func(t *testing.T) {
		if client.Projects() == nil {
			t.Fatal("expected projects resource")
		}
	})
}

func TestError(t *testing.T) {
	t.Run("creates error with properties", func(t *testing.T) {
		err := errors.New(
			"LUNA_ERR_TEST",
			"Test error",
			500,
			"req_123",
			nil,
		)

		if err.Code != "LUNA_ERR_TEST" {
			t.Errorf("expected code LUNA_ERR_TEST, got %s", err.Code)
		}
		if err.Message != "Test error" {
			t.Errorf("expected message 'Test error', got %s", err.Message)
		}
		if err.Status != 500 {
			t.Errorf("expected status 500, got %d", err.Status)
		}
		if err.RequestID != "req_123" {
			t.Errorf("expected request ID req_123, got %s", err.RequestID)
		}
	})

	t.Run("generates docs URL", func(t *testing.T) {
		err := errors.New("LUNA_ERR_TEST", "Test", 500, "req_123", nil)
		expected := "https://docs.eclipse.dev/luna/errors#LUNA_ERR_TEST"
		if err.DocsURL() != expected {
			t.Errorf("expected %s, got %s", expected, err.DocsURL())
		}
	})

	t.Run("identifies retryable errors", func(t *testing.T) {
		serverErr := errors.New(errors.CodeServerInternal, "Internal", 500, "req_123", nil)
		authErr := errors.New(errors.CodeAuthInvalidKey, "Invalid", 401, "req_123", nil)

		if !serverErr.Retryable() {
			t.Error("expected server error to be retryable")
		}
		if authErr.Retryable() {
			t.Error("expected auth error to not be retryable")
		}
	})
}

File: packages/typescript/tests/client.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { LunaClient } from '../src/client.js';
import { LunaError, AuthenticationError, NotFoundError } from '../src/errors/index.js';

describe('LunaClient', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    describe('constructor', () => {
        it('should create client with API key', () => {
            const client = new LunaClient({
                apiKey: 'lk_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
            });

            expect(client).toBeInstanceOf(LunaClient);
            expect(client.version).toBe('1.0.0');
        });

        it('should create client with access token', () => {
            const client = new LunaClient({
                accessToken: 'test-access-token',
            });

            expect(client).toBeInstanceOf(LunaClient);
        });

        it('should throw error when no auth provided', () => {
            expect(() => new LunaClient({})).toThrow('Either apiKey or accessToken must be provided');
        });

        it('should use custom base URL', () => {
            const client = new LunaClient({
                apiKey: 'lk_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
                baseUrl: 'https://api.staging.eclipse.dev',
            });

            expect(client).toBeInstanceOf(LunaClient);
        });
    });

    describe('resources', () => {
        it('should expose users resource', () => {
            const client = new LunaClient({
                apiKey: 'lk_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
            });

            expect(client.users).toBeDefined();
            expect(typeof client.users.list).toBe('function');
            expect(typeof client.users.get).toBe('function');
            expect(typeof client.users.create).toBe('function');
            expect(typeof client.users.update).toBe('function');
            expect(typeof client.users.delete).toBe('function');
        });

        it('should expose projects resource', () => {
            const client = new LunaClient({
                apiKey: 'lk_test_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
            });

            expect(client.projects).toBeDefined();
            expect(typeof client.projects.list).toBe('function');
            expect(typeof client.projects.get).toBe('function');
            expect(typeof client.projects.create).toBe('function');
            expect(typeof client.projects.update).toBe('function');
            expect(typeof client.projects.delete).toBe('function');
        });
    });
});

describe('LunaError', () => {
    it('should create error with all properties', () => {
        const error = new LunaError({
            code: 'LUNA_ERR_TEST',
            message: 'Test error',
            status: 500,
            requestId: 'req_123',
            details: { key: 'value' },
        });

        expect(error.code).toBe('LUNA_ERR_TEST');
        expect(error.message).toBe('Test error');
        expect(error.status).toBe(500);
        expect(error.requestId).toBe('req_123');
        expect(error.details).toEqual({ key: 'value' });
        expect(error.docsUrl).toBe('https://docs.eclipse.dev/luna/errors#LUNA_ERR_TEST');
    });

    it('should serialize to JSON', () => {
        const error = new LunaError({
            code: 'LUNA_ERR_TEST',
            message: 'Test error',
            status: 500,
            requestId: 'req_123',
        });

        const json = error.toJSON();
        expect(json.code).toBe('LUNA_ERR_TEST');
        expect(json.message).toBe('Test error');
        expect(json.request_id).toBe('req_123');
    });

    it('should identify retryable errors', () => {
        const serverError = new LunaError({
            code: 'LUNA_ERR_SERVER_INTERNAL',
            message: 'Internal error',
            status: 500,
            requestId: 'req_123',
        });

        const authError = new AuthenticationError({
            code: 'LUNA_ERR_AUTH_INVALID_KEY',
            message: 'Invalid key',
            requestId: 'req_123',
        });

        expect(serverError.retryable).toBe(true);
        expect(authError.retryable).toBe(false);
    });
});

describe('Error classes', () => {
    it('should create AuthenticationError with 401 status', () => {
        const error = new AuthenticationError({
            code: 'LUNA_ERR_AUTH_INVALID_KEY',
            message: 'Invalid API key',
            requestId: 'req_123',
        });

        expect(error.status).toBe(401);
        expect(error.name).toBe('AuthenticationError');
    });

    it('should create NotFoundError with 404 status', () => {
        const error = new NotFoundError({
            code: 'LUNA_ERR_RESOURCE_NOT_FOUND',
            message: 'User not found',
            requestId: 'req_123',
        });

        expect(error.status).toBe(404);
        expect(error.name).toBe('NotFoundError');
    });
});
